common/src/main/java/io/pravega/common/util/btree/EntryIterator.java

```diff
@@ -0,0 +1,207 @@
+import lombok.NonNull;
+import lombok.val;
+
+/**
+ * Iterator for keys in a BTreeIndex.
+ */
+class EntryIterator implements AsyncIterator<List<PageEntry>> {
+    //region Members
+
+    private static final ByteArrayComparator KEY_COMPARATOR = new ByteArrayComparator();
+    private final ByteArraySegment firstKey;
+    private final boolean firstKeyInclusive;
+    private final ByteArraySegment lastKey;
+    private final boolean lastKeyInclusive;
+    private final LocatePage locatePage;
+    private final Duration fetchTimeout;
+    private final AtomicBoolean finished;
+    private final PageCollection pageCollection;
+    private final AtomicReference<PageWrapper> lastPage;
+    private final AtomicInteger processedPageCount;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the EntryIterator class.
+     *
+     * @param firstKey          A ByteArraySegment indicating the first Key to iterate from.
+     * @param firstKeyInclusive If true, firstKey will be included in the iteration (provided it exists), otherwise it will
+     *                          be excluded.
+     * @param lastKey           A ByteArraySegment indicating the last Key to iterate to.
+     * @param lastKeyInclusive  If true, lastKey will be included in the iteration (provided it exists), otherwise it will
+     *                          be excluded.
+     * @param locatePage        A Function that can be used to locate a specific BTreePage.
+     * @param indexLength       The current index length.
+     * @param fetchTimeout      Timeout for each invocation of locatePage.
+     */
+    EntryIterator(@NonNull ByteArraySegment firstKey, boolean firstKeyInclusive, @NonNull ByteArraySegment lastKey, boolean lastKeyInclusive,
+                  @NonNull LocatePage locatePage, long indexLength, @NonNull Duration fetchTimeout) {
+        // First, verify correctness.
+        int c = KEY_COMPARATOR.compare(firstKey, lastKey);
+        if (firstKeyInclusive && lastKeyInclusive) {
+            Preconditions.checkArgument(c <= 0, "firstKey must be smaller than or equal to lastKey.");
+        } else {
+            Preconditions.checkArgument(c < 0, "firstKey must be smaller than lastKey.");
+        }
+
+        // firstKey and firstKeyInclusive will change as we make progress in our iteration.
+        this.firstKey = firstKey;
+        this.firstKeyInclusive = firstKeyInclusive;
+        this.lastKey = lastKey;
+        this.lastKeyInclusive = lastKeyInclusive;
+        this.locatePage = locatePage;
+        this.fetchTimeout = fetchTimeout;
+        this.pageCollection = new PageCollection(indexLength);
+        this.lastPage = new AtomicReference<>(null);
+        this.finished = new AtomicBoolean();
+        this.processedPageCount = new AtomicInteger();
+    }
+
+    //endregion
+
+    //region AsyncIterator Implementation
+
+    @Override
+    public CompletableFuture<List<PageEntry>> getNext() {
```

Comment:
It is not obvious what happens when two threads call this in parallel. The loop in  getNextLeafPage looks like it might have issues.

Reply:
Fixed. I am disallowing a second concurrent call until the first one finishes.